File: sdk/src/push.ts
import { OwnFirebaseClient } from './client';
import type { PushDeviceToken, PushTopic, PaginatedResponse } from './types';

export interface PushNotificationPayload {
  title: string;
  body: string;
  data?: Record<string, string>;
  icon?: string;
  badge?: string;
}

export interface PushNotificationRecord {
  id: string;
  title: string;
  body: string;
  status: string;
  sent_at: string;
  recipient_count: number;
}

export class PushSDK extends OwnFirebaseClient {
  // ─── Device Tokens ───────────────────────────────────────────────────────────

  async registerToken(
    token: string,
    platform: 'ios' | 'android' | 'web'
  ): Promise<PushDeviceToken> {
    return this.request('POST', this.projectUrl('push/tokens/'), {
      token,
      platform,
    });
  }

  async listTokens(): Promise<PaginatedResponse<PushDeviceToken>> {
    return this.request('GET', this.projectUrl('push/tokens/'));
  }

  async deleteToken(id: string): Promise<void> {
    return this.request('DELETE', this.projectUrl(`push/tokens/${id}/`));
  }

  // ─── Topics ───────────────────────────────────────────────────────────────────

  async listTopics(): Promise<PaginatedResponse<PushTopic>> {
    return this.request('GET', this.projectUrl('push/topics/'));
  }

  async createTopic(name: string): Promise<PushTopic> {
    return this.request('POST', this.projectUrl('push/topics/'), { name });
  }

  async subscribeTopic(topicId: string): Promise<{ detail: string }> {
    return this.request(
      'POST',
      this.projectUrl(`push/topics/${topicId}/`),
      { action: 'subscribe' }
    );
  }

  // ─── Send Notifications ──────────────────────────────────────────────────────

  async sendToDevice(
    tokenId: string,
    notification: PushNotificationPayload
  ): Promise<PushNotificationRecord> {
    return this.request('POST', this.projectUrl('push/notifications/'), {
      target_type: 'device',
      target_id: tokenId,
      ...notification,
    });
  }

  async sendToTopic(
    topicId: string,
    notification: PushNotificationPayload
  ): Promise<PushNotificationRecord> {
    return this.request('POST', this.projectUrl('push/notifications/'), {
      target_type: 'topic',
      target_id: topicId,
      ...notification,
    });
  }

  async listNotifications(): Promise<PaginatedResponse<PushNotificationRecord>> {
    return this.request('GET', this.projectUrl('push/notifications/'));
  }

  // ─── Campaigns ───────────────────────────────────────────────────────────────

  async listCampaigns(): Promise<PaginatedResponse<PushNotificationRecord>> {
    return this.request('GET', this.projectUrl('push/campaigns/'));
  }

  async createCampaign(
    notification: PushNotificationPayload & {
      scheduled_at?: string;
      audience?: Record<string, unknown>;
    }
  ): Promise<PushNotificationRecord> {
    return this.request('POST', this.projectUrl('push/campaigns/'), notification);
  }
}

File: sdk/python-sdk/ownfirebase/push.py
"""OwnFirebase Push SDK."""

from typing import Any, Dict

from .client import OwnFirebaseClient


class PushSDK(OwnFirebaseClient):
    """Push notifications service."""

    # ─── Device Tokens ───────────────────────────────────────────────────────────

    def register_token(self, token: str, platform: str) -> Dict[str, Any]:
        """Register a device token. platform is one of 'fcm', 'apns', 'web'."""
        return self.request(
            'POST',
            self.project_url('push/tokens/'),
            json_data={'token': token, 'platform': platform},
        )

    def list_tokens(self) -> Dict[str, Any]:
        """List registered device tokens. Returns a paginated response."""
        return self.request('GET', self.project_url('push/tokens/'))

    def delete_token(self, id: str) -> None:
        """Delete a registered device token."""
        return self.request('DELETE', self.project_url(f'push/tokens/{id}/'))

    # ─── Topics ───────────────────────────────────────────────────────────────────

    def list_topics(self) -> Dict[str, Any]:
        """List push topics. Returns a paginated response."""
        return self.request('GET', self.project_url('push/topics/'))

    def create_topic(self, name: str) -> Dict[str, Any]:
        """Create a new push topic."""
        return self.request('POST', self.project_url('push/topics/'), json_data={'name': name})

    def subscribe_topic(self, topic_id: str, device_token_id: str) -> Dict[str, Any]:
        """Subscribe a device token to a topic."""
        return self.request(
            'POST',
            self.project_url(f'push/topics/{topic_id}/subscribe/'),
            json_data={'device_token_id': device_token_id},
        )

    # ─── Send Notifications ──────────────────────────────────────────────────────

    def send_to_device(self, token_id: str, notification: Dict[str, Any]) -> Dict[str, Any]:
        """Send a push notification to a single device token."""
        return self.request(
            'POST',
            self.project_url('push/notifications/'),
            json_data={'device_token': token_id, **notification},
        )

    def send_to_topic(self, topic_id: str, notification: Dict[str, Any]) -> Dict[str, Any]:
        """Send a push notification to all subscribers of a topic."""
        return self.request(
            'POST',
            self.project_url('push/notifications/'),
            json_data={'topic': topic_id, **notification},
        )

    def list_notifications(self) -> Dict[str, Any]:
        """List sent push notifications. Returns a paginated response."""
        return self.request('GET', self.project_url('push/notifications/'))

    # ─── Campaigns ───────────────────────────────────────────────────────────────

    def list_campaigns(self) -> Dict[str, Any]:
        """List push campaigns. Returns a paginated response."""
        return self.request('GET', self.project_url('push/campaigns/'))

    def create_campaign(self, notification: Dict[str, Any]) -> Dict[str, Any]:
        """Create a push campaign. notification may include scheduled_at/audience."""
        return self.request(
            'POST', self.project_url('push/campaigns/'), json_data=notification
        )
